controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -0,0 +1,663 @@
+		}
+	}
+
+	// Handle orphaned volume cleanup
+	if existingAttachment != nil {
+		if err := r.cleanupOrphanedVolumes(ctx, pvcVolumes); err != nil {
+			ctx.Logger.Error(err, "Failed to cleanup orphaned volumes")
+			// Continue with processing
+		}
+	}
+
+	if len(pvcVolumes) == 0 {
+		// No PVC volumes to process
+		if existingAttachment != nil {
+			// TODO: AKP: Delete existing CnsNodeVmBatchAttachmet or
+			// CnsNodeVmAttachment resource(s) when no volumes.
+			ctx.Logger.Info("Delete existing CnsNodeVmBatchAttachment",
+				"attachment", existingAttachment.Name)
+		}
+		return nil
+	}
+
+	// Use robust error handling for volume processing
+	if err := r.processVolumesWithErrorHandling(ctx, pvcVolumes); err != nil {
+		ctx.Logger.Error(err, "Failed to process volumes with error handling")
+		// Fall back to standard processing
+	}
+
+	// Create or update batch attachment
+	if existingAttachment == nil {
+		return r.createBatchAttachment(ctx, pvcVolumes)
+	}
+
+	return r.updateBatchAttachment(ctx, existingAttachment, pvcVolumes)
+}
+
+func (r *Reconciler) createBatchAttachment(
+	ctx *pkgctx.VolumeContext,
+	volumes []vmopv1.VirtualMachineVolume) error {
+
+	attachmentName := pkgutil.CNSAttachmentNameForVolume(ctx.VM.Name, "batch")
+
+	volumeSpecs, err := r.BuildVolumeSpecs(ctx, volumes)
+	if err != nil {
+		return fmt.Errorf("failed to build volume specs: %w", err)
+	}
+
+	// Validate the volume specs before creating the batch attachment
+	if err := r.validateVolumeSpecs(ctx, volumeSpecs); err != nil {
+		return fmt.Errorf("volume spec validation failed: %w", err)
+	}
+
+	attachment := &cnsv1alpha1.CnsNodeVmBatchAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      attachmentName,
+			Namespace: ctx.VM.Namespace,
+		},
+		Spec: cnsv1alpha1.CnsNodeVmBatchAttachmentSpec{
+			NodeUUID: ctx.VM.Status.BiosUUID,
+			Volumes:  volumeSpecs,
+		},
+	}
+
+	if err := controllerutil.SetControllerReference(ctx.VM, attachment, r.Client.Scheme()); err != nil {
+		return fmt.Errorf("cannot set controller reference on CnsNodeVmBatchAttachment: %w", err)
+	}
+
+	if err := r.Create(ctx, attachment); err != nil {
+		return fmt.Errorf("cannot create CnsNodeVmBatchAttachment: %w", err)
+	}
+
+	ctx.Logger.Info("Created CnsNodeVmBatchAttachment", "attachment", attachmentName)
+	return nil
+}
+
+func (r *Reconciler) updateBatchAttachment(
+	ctx *pkgctx.VolumeContext,
+	existing *cnsv1alpha1.CnsNodeVmBatchAttachment,
+	volumes []vmopv1.VirtualMachineVolume) error {
+
+	volumeSpecs, err := r.BuildVolumeSpecs(ctx, volumes)
+	if err != nil {
+		return fmt.Errorf("failed to build volume specs: %w", err)
+	}
+
+	// Check if update is needed
+	if r.volumeSpecsEqual(existing.Spec.Volumes, volumeSpecs) {
+		// No update needed, just update VM status from existing attachment
+		return r.updateVMStatusFromBatchAttachment(ctx, existing)
+	}
+
+	// Update the batch attachment with new volume specs
+	existing.Spec.Volumes = volumeSpecs
```

Comment:
It should also validate quota and stuff as well? 
```
validateVolumeSpecs
```

Reply:
Hmm.  Remember that these are PVCs that already exist and we are only handling their attachment with the VM via this new resource.  CSI would have handled the quota evaluation and enforcement during volume creation.